AppDashboard/lib/app_dashboard_data.py

```diff
@@ -647,3 +729,125 @@ def can_upload_apps(self):
+        LOOKUP_DICT = self.build_dict(email=email)
+        temp_dict['nav'] = [{key: LOOKUP_DICT.get(key)} for key in
+                            temp_dict.get('nav') if
+                            key in LOOKUP_DICT]
+
+        temp_dict['panel'] = [{key: LOOKUP_DICT.get(key)} for key in
+                              temp_dict.get('panel') if
+                              key in LOOKUP_DICT]
+        user_info.dash_layout_settings = temp_dict
+        user_info.put()
+      return
+    except Exception as err:
+      logging.exception(err)
+      pass
+
+
+  def rebuild_dash_layout_settings_dict(self, email=None):
+    """ Rebuilds the user's layout settings in case there is an update
+      to the lookup dictionary
+
+    Args:
+      email: A str that indicates the e-mail address of the administrator of
+        this application, or None if the currently logged-in user is the admin.
+    """
+    if email is None:
+      user = users.get_current_user()
+      if not user:
+        return
+      email = user.email()
+    try:
+      user_info = self.get_by_id(UserInfo, email)
+      if user_info:
+        try:
+          if user_info.dash_layout_settings:
+            LOOKUP_DICT = self.build_dict(email=email)
+            values = user_info.dash_layout_settings
+            default_nav = ["app_management","appscale_management",
+                           "debugging_monitoring"]
+
+            nav_list=[]
+            for key_dict in values.get('nav'):
+              for temp_key in key_dict:
+                nav_list.append(temp_key)
+
+            print(nav_list)
+            if set(nav_list)!=set(default_nav):
+              for key in default_nav:
+                if nav_list.count(key)==0:
+                  nav_list.append(key)
+            values['nav'] = [{key: LOOKUP_DICT.get(key)} for key in
+                                nav_list if
+                                key in LOOKUP_DICT]
+            values['panel'] = [{key: LOOKUP_DICT.get(key)} for key_dict in
+                               values.get('panel') for key in key_dict if
+                               key in LOOKUP_DICT]
+            user_info.dash_layout_settings = values
+            user_info.put()
+          else:
+            self.set_dash_layout_settings()
+        except Exception as err:
+          logging.exception(err)
+          self.set_dash_layout_settings()
+    except Exception as err:
+      logging.exception(err)
+
+
+  def get_dash_layout_settings(self, email=None):
+    """ Queries the UserAppServer to see what settings the user has saved
+    for customizing the UI of the Dashboard.
+
+    Args:
+      email: A str that indicates the e-mail address of the administrator of
+        this application, or None if the currently logged-in user is the admin.
+
+    Returns:
+      A dictionary containing the customization layout
+    """
+    if email is None:
+      user = users.get_current_user()
+      if not user:
+        return
```

Comment:
Some places that call this function assume the output is a dictionary. Returning `None` will cause problems.